### PR TITLE
[Do not Review] Support hyper log log plus plus(HLL++)

### LINKED ISSUE #10652: [FEA] Support for approx_count_distinct
**Is your feature request related to a problem? Please describe.**
I would like to be able to implement a GPU version of Spark's `approx_count_distinct` function, which uses the [HyperLogLog++](https://en.wikipedia.org/wiki/HyperLogLog) cardinality estimation algorithm. 

cuDF does not appear to provide any features today that would allow me to do this.

**Describe the solution you'd like**
I would like cuDF to implement this capability and expose an API that is likely similar to `approx_percentile` in that there would be methods both for computing and merging the underlying data structure, whether that is based on HyperLogLog++ or some other algorithm.

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by res-life**
closes #10652 

- [x] Group HLL
- [x] Test: Group HLL
- [x] Compact plain sketch to Spark compatible
- [x] Test: Compact plain sketch to Spark compatible
- [x] Merge HLL
- [x] Test Merge HLL
- [x] Reduction

#### HLL++ description
First get a 64 bits hash code, and generate a interger pair: register index -> register value.  The register index is in [0, 512) if precision is 9.
Second, merge in the same group:
For each register index, find the max register value.
So HLL++ is actually like a process of doing max aggregation 512 times on 512 interger columns.

e.g., below is 2 sketches in a group.
|register index 0 | register index 1 | ... | register index 511|
|-|-|-|-|
|1|22|...|44|
|11|2|...|4|
Aggregation result is:

|register index 0 | register index 1 | ... | register index 511|
|-|-|-|-|
|11|22|...|44|

#### reduce_by_key memory issue
`thrust::reduce_by_key` will allocates `n` sketch values (n is the num_rows) which are large, typically each sketch is 512 integers(512 * 4 = 2K bytes). If there are 1G rows, it means `reduce_by_key` will use 2k * 1G = 2T intermidate memory.

[link](https://github.com/NVIDIA/thrust/blob/main/thrust/system/detail/generic/reduce_by_key.inl#L112)
```
    // scan the values by flag
    thrust::detail::temporary_array<ValueType,ExecutionPolicy> scanned_values(exec, n);
```
`thrust::reduce_by_key` is not suitable for merge HLL sketches.
This PR uses a new way to do aggregate.

#### aggregation steps this PR
Step1:
Partial merge:
Because group labels is sorted, so the group labels in each segment is also sorted.
Split rows into small segments with each segment contains 256 items.
One thread handles one segement.
In each segment, produces a cache recording the max values. It only uses N/256 caches which is relatively small memory. When scaning the items in a segment, when meets a new group, outputs the result for the previous group.

Step 2:
Merge the `cache` and the `result` to get the final result.

For more details, please refer to the comments in the code.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by vyasr** [comment]
Please link to https://github.com/rapidsai/cudf/issues/10652 as appropriate.

File: cpp/include/cudf/aggregation.hpp [hunk 1]
```diff
@@ -84,43 +84,45 @@ class aggregation {
...
+    MERGE_LISTS,      ///< merge multiple lists values into one list
+    MERGE_SETS,       ///< merge multiple lists values into one list then drop duplicate entries
+    MERGE_M2,         ///< merge partial values of M2 aggregation,
+    COVARIANCE,       ///< covariance between two sets of elements
+    CORRELATION,      ///< correlation between two sets of elements
+    TDIGEST,          ///< create a tdigest from a set of input values
+    MERGE_TDIGEST,    ///< create a tdigest by merging multiple tdigests together
+    HISTOGRAM,        ///< compute frequency of each element
+    MERGE_HISTOGRAM,  ///< merge partial values of HISTOGRAM aggregation
+    HLLPP,  ///< approximating the number of distinct items by using hyper log log plus plus (HLLPP)
```
Comment (ttnghia) [inline cpp/include/cudf/aggregation.hpp:124]:
```suggestion
    HLLPP,  ///< approximating the number of distinct items using HyperLogLogPlusPlus
```

File: cpp/include/cudf/hashing/detail/xxhash_64.cuh [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/pair.h>
+#include <thrust/tabulate.h>
+
+namespace cudf::hashing::detail {
+using hash_value_type = uint64_t;
+
+template <typename Key>
+struct XXHash_64 {
```
Comment (ttnghia) [inline cpp/include/cudf/hashing/detail/xxhash_64.cuh:36]: This seems to be moved from the source file, which was added in https://github.com/rapidsai/cudf/pull/13612. However, I'm confused with the other `XXHash_64` added in https://github.com/NVIDIA/spark-rapids-jni/pull/1248. What is the difference between them? Is this `XXHash_64` enough for addressing Spark's behavior in HLLPP?

File: cpp/src/groupby/sort/aggregate.cpp [hunk 3]
```diff
@@ -792,6 +808,24 @@ void aggregate_result_functor::operator()<aggregation::MERGE_TDIGEST>(aggregatio
...
 }
 
+template <>
+void aggregate_result_functor::operator()<aggregation::MERGE_HLLPP>(aggregation const& agg)
+{
+  if (cache.has_result(values, agg)) { return; }
+
+  int const precision =
+    dynamic_cast<cudf::detail::merge_hyper_log_log_aggregation const&>(agg).precision;
+
```
Comment (ttnghia) [inline cpp/src/groupby/sort/aggregate.cpp:818]:
```suggestion
  int const precision =
    dynamic_cast<cudf::detail::merge_hyper_log_log_aggregation const&>(agg).precision;
```

File: cpp/src/groupby/sort/group_hyper_log_log_plus_plus.cu [hunk 4]
```diff
@@ -0,0 +1,674 @@
...
+#include <cudf/hashing/detail/xxhash_64.cuh>
+#include <cudf/structs/structs_column_view.hpp>
+#include <cudf/utilities/memory_resource.hpp>
+#include <cudf/utilities/span.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <cuco/hyperloglog_ref.cuh>
```
Comment (ttnghia) [inline cpp/src/groupby/sort/group_hyper_log_log_plus_plus.cu:34]: Unused?

File: cpp/src/groupby/sort/aggregate.cpp (hunk 3 above)
Comment (mythrocks) [reply cpp/src/groupby/sort/aggregate.cpp:818]: Sorry, my eyes aren't able to pick up the difference suggested here.  What's the change?

Comment (wence-) [reply cpp/src/groupby/sort/aggregate.cpp:818]: I suspect removal of the blank line, but github has a bug that doesn't show it.

File: cpp/src/groupby/sort/group_hyper_log_log_plus_plus.cu (hunk 4 above)
Comment (res-life) [reply cpp/src/groupby/sort/group_hyper_log_log_plus_plus.cu]: yes. done.

File: cpp/include/cudf/aggregation.hpp (hunk 1 above)
Comment (res-life) [reply cpp/include/cudf/aggregation.hpp]: done.

**Comment by res-life** [comment]
Ready to review except test cases.

File: cpp/src/groupby/sort/group_reductions.hpp [hunk 5]
```diff
@@ -539,6 +539,21 @@ std::unique_ptr<column> group_correlation(column_view const& covariance,
                                           rmm::cuda_stream_view stream,
                                           rmm::device_async_resource_ref mr);
 
+std::unique_ptr<column> group_hyper_log_log_plus_plus(
+  column_view const& input,
+  int64_t const num_groups,
+  cudf::device_span<size_type const> group_lables,
```
Comment (ustcfy) [inline cpp/src/groupby/sort/group_reductions.hpp:545]: Nit: `lable` --> `label`

File: cpp/include/cudf/detail/hyper_log_log_plus_plus/hyper_log_log_plus_plus.hpp [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+#include <cudf/types.hpp>
+#include <cudf/utilities/memory_resource.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+
+namespace cudf {
+namespace groupby::detail {
+
+/**
+ * Compute the hashs of the input column, then generate a scalar that is a sketch in long array
```
Comment (ustcfy) [inline cpp/include/cudf/detail/hyper_log_log_plus_plus/hyper_log_log_plus_plus.hpp:28]: Nit: `hashs` --> `hashes`

**Comment by res-life** [comment]
Moved to https://github.com/NVIDIA/spark-rapids-jni/pull/2522 using Host UDF approach, close this.

**Pull request closed without merging**